worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionFilesSorter.java

```diff
@@ -629,10 +629,18 @@ public void sort() throws InterruptedException {
           for (ShuffleBlockInfo blockInfo : originShuffleBlocks) {
             long offset = blockInfo.offset;
             long length = blockInfo.length;
-            ShuffleBlockInfo sortedBlock = new ShuffleBlockInfo();
-            sortedBlock.offset = fileIndex;
-            sortedBlock.length = length;
-            sortedShuffleBlocks.add(sortedBlock);
+            // combine multiple small length `ShuffleBlockInfo` for same mapId such that
+            // size of compacted `ShuffleBlockInfo` does not exceed `shuffleChunkSize`
+            if (!sortedShuffleBlocks.isEmpty()
+                && sortedShuffleBlocks.get(sortedShuffleBlocks.size() - 1).length + length
+                    <= shuffleChunkSize) {
```

Comment:
The condition "<= shuffleChunkSize" may cause the returned chunk's size to approach 2 * chunkSize. This won't be a trouble but changes the default behavior.

Reply:
Yes, say `shuffleChunkSize` is 8m, the generated blocks can be two 7.9m chunks, when fetch data it can return the two chunks together, which is 15.8m. That's the reason why I mentioned in the jira to use half of `shuffleChunkSize`, in which the worst case is 3.9m * 3 = 12.7m. So multiply by a fraction can control the exceeded size, maybe 0.25 is better (upper bound for default config is 8m + 0.25 * 8m = 10m)

Reply:
yeah, that sounds good. I can make this threshold configurable with default value of 0.25 to start with.

Although i'm not super clear about the issue in the discussion. Why would fetch data will return two or three chunks together making it 7.9 * 2 = 15.8m or 3.9 * 3 = 11.7 as mentioned above. Do clients cache the initial read chunk while reading the next one. If that is the case, can someone point me to this piece of code.

Reply:
> yeah, that sounds good. I can make this threshold configurable with default value of 0.25 to start with.
> 
> Although i'm not super clear about the issue in the discussion. Why would fetch data will return two or three chunks together making it 7.9 * 2 = 15.8m or 3.9 * 3 = 11.7 as mentioned above. Do clients cache the initial read chunk while reading the next one. If that is the case, can someone point me to this piece of code.

Hi @s0nskar , when creating `ReduceFileMeta` it will compact contiguous `ShuffleBlockInfo`s into a chunk, it first adds the current `ShuffleBlockInfo` then checks whether size exceeds `fetchChunkSize`:

https://github.com/apache/celeborn/blob/ba499704d1b058000ac20cffda93a6a39578eabe/common/src/main/java/org/apache/celeborn/common/util/ShuffleBlockInfoUtils.java#L53

Reply:
got it, thanks for the context.

Reply:
As per documentation `fetchChunkSize` is "Max chunk size". So shouldn't we change above condition to respect it. As you mentioned, if the ShuffleBlocks are generated close of 8mb then chunk can be close to 16mb.

We can change this condition to generate offsets closer to fetchChunkSize with some error factor of 10 or 20%. wdyt?